### PR TITLE
[2788] Add the possibility to prevent node to be draggable

### LINKED ISSUE #2788: Add the ability to stop considering some nodes as draggable
- Given a diagram with a node using a list layout with a couple of child nodes also using list layout (think classes in Papyrus Web using a list layout with the properties child node also using a list layout)
- When I try to perform a drag operation by clicking on the child node (not the grand-children like a property but the node containing all the properties)
- Then the node should not be draggable because it makes no sense in my use case.

The best scenario here for me should be that the parent node should be dragged instead but it may not be achievable...

### PR DESCRIPTION
**Pull request by frouene**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2788

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java [hunk 1]
```diff
@@ -127,21 +127,21 @@ public Optional<IRepresentationDescription> findById(IEditingContext editingCont
...
                 .to(SECOND_TARGET_OBJECT_ID).at(0.25, 0.5)
                 .and()
-            .edge("two")
+                .edge("two")
                 .from(FIRST_TARGET_OBJECT_ID).at(0.5, 0.75)
                 .to(THIRD_TARGET_OBJECT_ID).at(0.25, 0.5)
                 .goingThrough(20, 50)
                 .and()
-            .build();
+                .build();
```
Comment (pcdavid) [inline packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java]: I think the previous formatting was on purpose, to clearly identify "sub blocks" corresponding to the different elements being built.

File: packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java [hunk 2]
```diff
@@ -183,24 +183,24 @@ public void testSimpleDiagramLayout() throws IOException {
...
                 .to(secondParentTargetObjectId).at(0.8, 0.8)
                 .and()
-            .edge("Opposite Link")
+                .edge("Opposite Link")
                 .from(secondParentTargetObjectId).at(0.2, 0.7)
                 .to(firstParentTargetObjectId).at(0.1, 0.9)
-            .and()
-            .build();
+                .and()
+                .build();
```
Comment (pcdavid) [inline packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java]: Same remark.

File: packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java [hunk 3]
```diff
@@ -250,28 +250,28 @@ public void testSimpleDiagramLayout() throws IOException {
...
+                .of(-1, -1)
+                .childNodes(columnListLayoutStrategy).and()
+                .and()
+                .and()
+                .edge(SECOND_TARGET_OBJECT_ID)
                 .from(FIRST_TARGET_OBJECT_ID).at(0.75, 0.25)
                 .to(SECOND_TARGET_OBJECT_ID).at(0.25, 0.25)
                 .and()
-            .build();
+                .build();
```
Comment (pcdavid) [inline packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java]: Same remark.

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/diagram/ViewDiagramDescriptionConverter.java [hunk 4]
```diff
@@ -231,8 +231,9 @@ private NodeDescription convert(org.eclipse.sirius.components.view.diagram.NodeD
             ILayoutStrategy childrenLayoutStrategy = null;
 
             LayoutStrategyDescription childrenLayoutStrategyFromViewModel = viewNodeDescription.getChildrenLayoutStrategy();
-            if (childrenLayoutStrategyFromViewModel instanceof ListLayoutStrategyDescription) {
-                childrenLayoutStrategy = new ListLayoutStrategy();
+            if (childrenLayoutStrategyFromViewModel instanceof ListLayoutStrategyDescription listLayoutStrategyDescription) {
+                Result result = interpreter.evaluateExpression(variableManager.getVariables(), listLayoutStrategyDescription.getAreChildNodesDraggableExpression());
+                childrenLayoutStrategy = ListLayoutStrategy.newListLayoutStrategy().areChildNodesDraggable(result.asBoolean().orElse(false)).build();
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/diagram/ViewDiagramDescriptionConverter.java]: `ListLayoutStrategy` has `true` as the default value, but here we set `false` if the expression is not set/fails during its evaluation. Is it voluntary?

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/diagram/ViewDiagramDescriptionConverter.java (hunk 4 above)
Comment (frouene) [reply packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/diagram/ViewDiagramDescriptionConverter.java]: Yes, you're right the default value is true

File: packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java (hunk 2 above)
Comment (frouene) [reply packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java]: done

File: packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java (hunk 3 above)
Comment (frouene) [reply packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java]: done

File: packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java (hunk 1 above)
Comment (frouene) [reply packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTests.java]: I keep the old formatting

**Review by pcdavid — APPROVED** [review]

**Comment by gcoutable** [comment]
The failing has nothing to do with this PR, I will merge it

**Pull request merged**